File: src/main/db/schema.ts
import { integer, sqliteTable, text } from 'drizzle-orm/sqlite-core'

export const sessions = sqliteTable('sessions', {
  id: text('id').primaryKey(),
  title: text('title').notNull(),
  testTarget: text('test_target'),
  charter: text('charter'),
  environment: text('environment'),
  buildVersion: text('build_version'),
  relatedReference: text('related_reference'),
  createdAt: text('created_at').notNull(),
  updatedAt: text('updated_at').notNull(),
  lastOpenedAt: text('last_opened_at').notNull()
})

export const entries = sqliteTable('entries', {
  id: text('id').primaryKey(),
  sessionId: text('session_id')
    .notNull()
    .references(() => sessions.id, { onDelete: 'cascade' }),
  type: text('type', {
    enum: ['note', 'observation', 'api_response', 'log', 'screenshot', 'finding_candidate']
  }).notNull(),
  title: text('title'),
  body: text('body').notNull(),
  metadataJson: text('metadata_json'),
  createdAt: text('created_at').notNull(),
  updatedAt: text('updated_at').notNull(),
  excludedFromGeneration: integer('excluded_from_generation', { mode: 'boolean' }).notNull().default(false)
})

export const attachments = sqliteTable('attachments', {
  id: text('id').primaryKey(),
  sessionId: text('session_id')
    .notNull()
    .references(() => sessions.id, { onDelete: 'cascade' }),
  entryId: text('entry_id').references(() => entries.id, { onDelete: 'set null' }),
  filename: text('filename').notNull(),
  mimeType: text('mime_type'),
  sizeBytes: integer('size_bytes').notNull(),
  sha256: text('sha256').notNull(),
  relativePath: text('relative_path').notNull(),
  createdAt: text('created_at').notNull()
})

export const findings = sqliteTable('findings', {
  id: text('id').primaryKey(),
  sessionId: text('session_id')
    .notNull()
    .references(() => sessions.id, { onDelete: 'cascade' }),
  title: text('title').notNull(),
  body: text('body').notNull(),
  kind: text('kind', { enum: ['bug', 'question', 'risk', 'follow_up', 'note'] }).notNull(),
  metadataJson: text('metadata_json'),
  createdAt: text('created_at').notNull(),
  updatedAt: text('updated_at').notNull()
})

export const evidenceLinks = sqliteTable('evidence_links', {
  id: text('id').primaryKey(),
  findingId: text('finding_id')
    .notNull()
    .references(() => findings.id, { onDelete: 'cascade' }),
  entryId: text('entry_id').references(() => entries.id, { onDelete: 'cascade' }),
  attachmentId: text('attachment_id').references(() => attachments.id, { onDelete: 'cascade' }),
  createdAt: text('created_at').notNull()
})

export const generationContexts = sqliteTable('generation_contexts', {
  id: text('id').primaryKey(),
  sessionId: text('session_id')
    .notNull()
    .references(() => sessions.id, { onDelete: 'cascade' }),
  createdAt: text('created_at').notNull()
})

export const generationContextEntries = sqliteTable('generation_context_entries', {
  id: text('id').primaryKey(),
  generationContextId: text('generation_context_id')
    .notNull()
    .references(() => generationContexts.id, { onDelete: 'cascade' }),
  entryId: text('entry_id')
    .notNull()
    .references(() => entries.id, { onDelete: 'cascade' }),
  included: integer('included', { mode: 'boolean' }).notNull().default(true)
})

export const generationContextAttachments = sqliteTable('generation_context_attachments', {
  id: text('id').primaryKey(),
  generationContextId: text('generation_context_id')
    .notNull()
    .references(() => generationContexts.id, { onDelete: 'cascade' }),
  attachmentId: text('attachment_id')
    .notNull()
    .references(() => attachments.id, { onDelete: 'cascade' }),
  included: integer('included', { mode: 'boolean' }).notNull().default(true)
})

export const aiRuns = sqliteTable('ai_runs', {
  id: text('id').primaryKey(),
  sessionId: text('session_id')
    .notNull()
    .references(() => sessions.id, { onDelete: 'cascade' }),
  generationContextId: text('generation_context_id').references(() => generationContexts.id, { onDelete: 'set null' }),
  provider: text('provider', {
    enum: ['apple_intelligence', 'claude_code', 'codex_cli', 'copilot_cli']
  }).notNull(),
  model: text('model').notNull(),
  reasoningEffort: text('reasoning_effort', { enum: ['minimal', 'low', 'medium', 'high', 'xhigh', 'max'] }),
  promptVersion: text('prompt_version').notNull(),
  status: text('status', { enum: ['running', 'completed', 'failed'] }).notNull(),
  errorMessage: text('error_message'),
  createdAt: text('created_at').notNull(),
  completedAt: text('completed_at')
})

export const drafts = sqliteTable('drafts', {
  id: text('id').primaryKey(),
  sessionId: text('session_id')
    .notNull()
    .references(() => sessions.id, { onDelete: 'cascade' }),
  aiRunId: text('ai_run_id').references(() => aiRuns.id, { onDelete: 'set null' }),
  kind: text('kind', { enum: ['session_report'] }).notNull(),
  title: text('title').notNull(),
  body: text('body').notNull(),
  createdAt: text('created_at').notNull(),
  updatedAt: text('updated_at').notNull()
})

File: crates/qa-scribe-core/src/storage/schema.rs
pub(super) const SCHEMA: &str = r#"
CREATE TABLE IF NOT EXISTS sessions (
  id TEXT PRIMARY KEY,
  title TEXT NOT NULL CHECK (length(trim(title)) > 0),
  session_context TEXT,
  objective_notes TEXT,
  environment TEXT,
  build_version TEXT,
  related_reference TEXT,
  created_at TEXT NOT NULL,
  updated_at TEXT NOT NULL,
  last_opened_at TEXT NOT NULL
);

CREATE TABLE IF NOT EXISTS app_settings (
  key TEXT PRIMARY KEY,
  value_json TEXT NOT NULL,
  updated_at TEXT NOT NULL
);

CREATE TABLE IF NOT EXISTS entries (
  id TEXT PRIMARY KEY,
  session_id TEXT NOT NULL REFERENCES sessions(id) ON DELETE CASCADE,
  type TEXT NOT NULL CHECK (type IN ('note', 'observation', 'api_response', 'log', 'screenshot', 'finding_candidate')),
  title TEXT,
  body TEXT NOT NULL,
  body_json TEXT,
  body_format TEXT NOT NULL DEFAULT 'html',
  metadata_json TEXT,
  excluded_from_generation INTEGER NOT NULL DEFAULT 0,
  created_at TEXT NOT NULL,
  updated_at TEXT NOT NULL
);

CREATE TABLE IF NOT EXISTS attachments (
  id TEXT PRIMARY KEY,
  session_id TEXT NOT NULL REFERENCES sessions(id) ON DELETE CASCADE,
  entry_id TEXT REFERENCES entries(id) ON DELETE SET NULL,
  filename TEXT NOT NULL,
  mime_type TEXT,
  size_bytes INTEGER NOT NULL CHECK (size_bytes >= 0),
  sha256 TEXT NOT NULL,
  relative_path TEXT NOT NULL,
  created_at TEXT NOT NULL
);

CREATE TABLE IF NOT EXISTS findings (
  id TEXT PRIMARY KEY,
  session_id TEXT NOT NULL REFERENCES sessions(id) ON DELETE CASCADE,
  title TEXT NOT NULL CHECK (length(trim(title)) > 0),
  body TEXT NOT NULL,
  body_json TEXT,
  body_format TEXT NOT NULL DEFAULT 'html',
  kind TEXT NOT NULL CHECK (kind IN ('bug', 'question', 'risk', 'follow_up', 'note')),
  metadata_json TEXT,
  created_at TEXT NOT NULL,
  updated_at TEXT NOT NULL
);

CREATE TABLE IF NOT EXISTS evidence_links (
  id TEXT PRIMARY KEY,
  finding_id TEXT NOT NULL REFERENCES findings(id) ON DELETE CASCADE,
  entry_id TEXT REFERENCES entries(id) ON DELETE CASCADE,
  attachment_id TEXT REFERENCES attachments(id) ON DELETE CASCADE,
  created_at TEXT NOT NULL,
  CHECK (entry_id IS NOT NULL OR attachment_id IS NOT NULL)
);

CREATE TABLE IF NOT EXISTS generation_contexts (
  id TEXT PRIMARY KEY,
  session_id TEXT NOT NULL REFERENCES sessions(id) ON DELETE CASCADE,
  created_at TEXT NOT NULL
);

CREATE TABLE IF NOT EXISTS generation_context_entries (
  id TEXT PRIMARY KEY,
  generation_context_id TEXT NOT NULL REFERENCES generation_contexts(id) ON DELETE CASCADE,
  entry_id TEXT NOT NULL REFERENCES entries(id) ON DELETE CASCADE,
  included INTEGER NOT NULL DEFAULT 1
);

CREATE TABLE IF NOT EXISTS generation_context_attachments (
  id TEXT PRIMARY KEY,
  generation_context_id TEXT NOT NULL REFERENCES generation_contexts(id) ON DELETE CASCADE,
  attachment_id TEXT NOT NULL REFERENCES attachments(id) ON DELETE CASCADE,
  included INTEGER NOT NULL DEFAULT 1
);

CREATE TABLE IF NOT EXISTS ai_runs (
  id TEXT PRIMARY KEY,
  session_id TEXT NOT NULL REFERENCES sessions(id) ON DELETE CASCADE,
  generation_context_id TEXT REFERENCES generation_contexts(id) ON DELETE SET NULL,
  provider TEXT NOT NULL CHECK (provider IN ('claude_code', 'codex_cli', 'copilot_cli')),
  model TEXT NOT NULL,
  reasoning_effort TEXT,
  prompt_version TEXT NOT NULL,
  status TEXT NOT NULL CHECK (status IN ('running', 'completed', 'failed')),
  error_message TEXT,
  created_at TEXT NOT NULL,
  completed_at TEXT
);

CREATE TABLE IF NOT EXISTS drafts (
  id TEXT PRIMARY KEY,
  session_id TEXT NOT NULL REFERENCES sessions(id) ON DELETE CASCADE,
  ai_run_id TEXT REFERENCES ai_runs(id) ON DELETE SET NULL,
  kind TEXT NOT NULL CHECK (kind IN ('session_report', 'testware')),
  title TEXT NOT NULL CHECK (length(trim(title)) > 0),
  body TEXT NOT NULL,
  body_json TEXT,
  body_format TEXT NOT NULL DEFAULT 'html',
  metadata_json TEXT,
  created_at TEXT NOT NULL,
  updated_at TEXT NOT NULL
);

CREATE INDEX IF NOT EXISTS idx_sessions_last_opened ON sessions(last_opened_at DESC);
CREATE INDEX IF NOT EXISTS idx_entries_session_created ON entries(session_id, created_at);
CREATE INDEX IF NOT EXISTS idx_attachments_session_created ON attachments(session_id, created_at);
CREATE INDEX IF NOT EXISTS idx_findings_session_created ON findings(session_id, created_at);
CREATE INDEX IF NOT EXISTS idx_drafts_session_updated ON drafts(session_id, updated_at);
CREATE INDEX IF NOT EXISTS idx_ai_runs_session_created ON ai_runs(session_id, created_at);
CREATE INDEX IF NOT EXISTS idx_evidence_links_finding_id ON evidence_links(finding_id);
CREATE INDEX IF NOT EXISTS idx_evidence_links_entry_id ON evidence_links(entry_id);
CREATE INDEX IF NOT EXISTS idx_evidence_links_attachment_id ON evidence_links(attachment_id);
CREATE INDEX IF NOT EXISTS idx_generation_context_entries_generation_context_id ON generation_context_entries(generation_context_id);
CREATE INDEX IF NOT EXISTS idx_generation_context_entries_entry_id ON generation_context_entries(entry_id);
CREATE INDEX IF NOT EXISTS idx_generation_context_attachments_generation_context_id ON generation_context_attachments(generation_context_id);
CREATE INDEX IF NOT EXISTS idx_generation_context_attachments_attachment_id ON generation_context_attachments(attachment_id);
CREATE INDEX IF NOT EXISTS idx_drafts_ai_run_id ON drafts(ai_run_id);
CREATE INDEX IF NOT EXISTS idx_attachments_entry_id ON attachments(entry_id);
CREATE INDEX IF NOT EXISTS idx_generation_contexts_session_id ON generation_contexts(session_id);
CREATE INDEX IF NOT EXISTS idx_ai_runs_generation_context_id ON ai_runs(generation_context_id);
CREATE INDEX IF NOT EXISTS idx_ai_runs_running_status ON ai_runs(status) WHERE status = 'running';
"#;
